File: src/lib/bridge/vault.ts
import { invoke } from '@tauri-apps/api/core';
import type { VaultStatus } from './type';

/**
 * @description 获取 Vault 状态
 * @returns Vault 状态
 */
export async function getVaultStatus() {
  return await invoke<VaultStatus>('vault_get_status');
}

/**
 * @description 解锁 Vault
 * @param password - Vault 密码
 * @returns Vault 状态
 */
export async function unlockVault(password: string) {
  return await invoke<VaultStatus>('vault_unlock', { password });
}

/**
 * @description 锁定 Vault
 * @returns Vault 状态
 */
export async function lockVault() {
  return await invoke<VaultStatus>('vault_lock');
}

/**
 * @description 设置 Vault 密码
 * @param password - Vault 密码
 * @returns Vault 状态
 */
export async function setupVaultPassword(password: string) {
  return await invoke<VaultStatus>('vault_setup_password', { password });
}

/**
 * @description 修改 Vault 密码
 * @param oldPassword - 旧密码
 * @param newPassword - 新密码
 */
export async function changeVaultPassword(oldPassword: string, newPassword: string) {
  return await invoke<void>('vault_change_password', { old: oldPassword, new: newPassword });
}

File: src-tauri/src/server/commands/vault.rs
use tauri::State;

use crate::server::dto::VaultStatusDto;
use crate::server::handlers;
use crate::server::state::AppState;
use crate::server::transport::to_command_result;

/// 获取 Vault 状态
/// 
/// @return  Vault 状态
#[tauri::command]
pub async fn vault_get_status(state: State<'_, AppState>) -> Result<VaultStatusDto, String> {
  let vault = state.vault.lock().await;
  to_command_result(handlers::vault::get_status(&state.db, &vault).await)
}

/// 解锁 Vault
/// 
/// @param password - 密码
/// 
/// @return  Vault 状态
#[tauri::command]
pub async fn vault_unlock(state: State<'_, AppState>, password: String) -> Result<VaultStatusDto, String> {
  let mut vault = state.vault.lock().await;
  to_command_result(handlers::vault::unlock(&state.db, &mut vault, password).await)
}

/// 锁定 Vault
/// 
/// @return  Vault 状态
#[tauri::command]
pub async fn vault_lock(state: State<'_, AppState>) -> Result<VaultStatusDto, String> {
  let mut vault = state.vault.lock().await;
  to_command_result(handlers::vault::lock(&state.db, &mut vault).await)
}

/// 设置 Vault 密码
///
/// @param password - 密码
/// 
/// @return  Vault 状态
#[tauri::command]
pub async fn vault_setup_password(state: State<'_, AppState>, password: String) -> Result<VaultStatusDto, String> {
  let mut vault = state.vault.lock().await;
  to_command_result(handlers::vault::setup_password(&state.db, &mut vault, password).await)
}

/// 修改 Vault 密码
/// 
/// @param state - 应用状态
/// @param old - 旧密码
/// @param new - 新密码
/// 
/// @return 结果
#[tauri::command]
pub async fn vault_change_password(state: State<'_, AppState>, old: String, new: String) -> Result<(), String> {
  let vault = state.vault.lock().await;
  to_command_result(handlers::vault::change_password(&state.db, &vault, old, new).await)
}
